src/cancel_stream.rs

```diff
@@ -0,0 +1,293 @@
+use borsh::BorshSerialize;
+use solana_program::{
+    borsh as solana_borsh,
+    entrypoint::ProgramResult,
+    msg,
+    program::{invoke, invoke_signed},
+    program_error::ProgramError,
+    pubkey::Pubkey,
+    sysvar::{clock::Clock, Sysvar},
+};
+use spl_associated_token_account::create_associated_token_account;
+
+use crate::{
+    error::SfError,
+    state::{InstructionAccounts, TokenStreamData},
+    stream_safety::{initialized_account_sanity_check, metadata_sanity_check},
+    utils::{calculate_available, encode_base10, unpack_mint_account},
+};
+
+/// Cancel an SPL Token stream
+///
+/// The function will read the instructions from the metadata account and see
+/// if there are any unlocked funds. If so, they will be transferred to the
+/// stream recipient.
+pub(crate) fn cancel(program_id: &Pubkey, acc: InstructionAccounts) -> ProgramResult {
+    msg!("Cancelling SPL token stream");
```

Comment:
refresh_amounts()
as first line in cancel, withdraw, topup
to read the actual state from the chain (i.e. to see if additional funds have been deposited)
```
metadata.deposited_amount = escrow_tok_acc.amount + metadata.withdrawn_amount;
if (old_deposited_amount != metadata.deposited_amount) {
  //update fees
}
```